[24.2] Ensure job states are fetched in invocation view

ahmedhamidawan opened the pull request:
Fixes https://github.com/galaxyproject/galaxy/issues/20003

We weren't checking if the job states summary had been fetched before we had the following check:
```
if (invocationSchedulingTerminal.value && jobCount.value === 0)
```
`jobCount` is always 0 if the states summary hasn't been fetched yet or has `states: {}`; which can happen when the `populated_state === "new"`. Now we check for that and keep polling until we see some other state than `new`.

### (Additional Improvement) Route to Invocation on Run
**For a singular invocation** (excluding cases of batch runs), we now push to the `workflows/invocations/{invocation_id}` route when the workflow is run.
Without this, we would stay on the `workflows/run` route and refreshing would take you back to the run form, instead of the invocation view.

### ~~TODO~~ Done ✅:
- [x] Add tests
(_If problematic (for merge forwards) I can target dev in another PR just for tests? Doesn't seem bad though..._)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
[24.2] Ensure job states are fetched in invocation view

Fixes https://github.com/galaxyproject/galaxy/issues/20003

We weren't checking if the job states summary had been fetched before we had the
```
if (invocationSchedulingTerminal.value && jobCount.value === 0)
```
check. `jobCount` would always be 0 if the states summary hadn't been fetched yet.

ahmedhamidawan committed:
fix `jobStatesSummary` return type (should be allowed to return `null`)

ahmedhamidawan committed:
route to invocation on run

**For a singular invocation** (excluding cases of batch runs), we now push to the `workflows/invocations/{invocation_id}` route when the workflow is run.
Without this, we would stay on the `workflows/run` route and refreshing would take you back to the run form, instead of the invocation view.

ahmedhamidawan commented on client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue:
```diff
@@ -108,18 +108,22 @@ const invocationSchedulingTerminal = computed(() => {
     if (invocationSchedulingTerminal.value && jobCount.value === 0) {
-        // no jobs for this invocation (think subworkflow or just inputs)
+        // no jobs for this invocation (think it has just subworkflows/inputs)
         return true;
```
This isn't fixing it entirely unfortunately.

We still get a case where **sometimes**, we fetch the job states summary just in time, that it is:
```
{ "id": "id_here", "states": {}, "populated_state": "new", "model": "WorkflowInvocation" } 
```
instead of
```
{ "id": "id_here", "states": { "queued": 2 }, "populated_state": "new", "model": "WorkflowInvocation" } 
```
And before the summary can get non terminal states, like so ^, `invocationSchedulingTerminal` becomes `true`, and so, the `jobCount` is considered 0, and we stop fetching the job summaries; even though we really need to since we see 2 steps with 2 expected jobs, like:

![image](https://github.com/user-attachments/assets/8b653a33-f8ec-4c0c-81b4-3e2fc6701d1b)

ahmedhamidawan replied:
So this next commit https://github.com/galaxyproject/galaxy/pull/20010/commits/642ad39a5dde2c4ee831e07e5daf179d564ca981 should fix it if I got the concept right:

We see the `populated_state` being set in `summarize_invocation_jobs` here https://github.com/galaxyproject/galaxy/blob/4a0cca07d97d7bbd22eb39ddab38d4d4354aa4fa/lib/galaxy/managers/jobs.py#L957-L1000, `new` seems to be a non-terminal `populated_state`, meaning this summary will have updates which will eventually produce some state counts.

To fix the bug of never getting a state count for the invocation view before the scheduling state becomes terminal, we now wait for the state to be something else than `new`, and only then look for other reasons to stop polling.

ahmedhamidawan commented on client/src/components/WorkflowInvocationState/WorkflowInvocationState.vue:
```diff
@@ -108,18 +108,22 @@ const invocationSchedulingTerminal = computed(() => {
+    if (jobStatesSummary.value === null) {
+        return false;
+    }
+
```
Nevermind, I have a solution that might be better

ahmedhamidawan committed:
use `populated_state` in the jobs summary to continue polling for states

As we can see this value being set in `summarize_invocation_jobs` here https://github.com/galaxyproject/galaxy/blob/4a0cca07d97d7bbd22eb39ddab38d4d4354aa4fa/lib/galaxy/managers/jobs.py#L957-L1000, `new` seems to be a non-terminal `populated_state`, meaning this summary will have updates which will eventually produce some state counts.

To fix the bug of never getting a state count for the invocation view before the scheduling state becomes terminal, we now wait for the state to be something else than `new`, and only then look for other reasons to stop polling.

ahmedhamidawan committed:
add jest to test `new` `populated_state` for job summary

mvdbeek reviewed (approved):
That looks good to me, thanks @ahmedhamidawan!

mvdbeek commented:
But the selenium tests look related ...

mvdbeek commented:
From the failing test:
![last](https://github.com/user-attachments/assets/3dfce373-73a5-40a7-881a-282ccca96f03)

I think there's a badge(?) missing here that let's you switch to the new history ?

ahmedhamidawan committed:
remove `newHistoryTarget` prop and instead check if history is different

This means that we show the little icon to switch to the history for the invocation if that history isn't current, regardless of if results were generated in a new history or not.

ahmedhamidawan commented:
This was happening because we now push to invocation route on workflow run, and hence, we were unable to pass the `newHistoryTarget` prop and show the badge.
Removed that prop and now we show the badge to switch to the history for the invocation if that history isn't current, regardless of if results were generated in a new history or not.

mvdbeek commented:
Excellent, that was another minor thing I didn't like, a 2 for 1!

mvdbeek commented:
Thank you!
